[draft-export] Include guards for constraint violation errors

angelayi opened the pull request:
Summary:
Added where logs are being added to constrain violations in draft export.

Example output:
```
1. Constraint violation error.
    The specified input dynamic_shapes spec was found to be incorrect during tracing.
    Specifically, this guard was added: Eq(s0, 3), where {'s0': "L['args'][0][0].size()[0]"}.
    This occured at the following stacktrace:
        File /data/users/angelayi/fbsource/buck-out/v2/gen/fbcode/1beb9df83fd74b9a/scripts/angelayi/draft_export/__test_draft_export__/test_draft_export#link-tree/torch/nn/modules/module.py, lineno 1736, in _wrapped_call_impl
        File /data/users/angelayi/fbsource/buck-out/v2/gen/fbcode/1beb9df83fd74b9a/scripts/angelayi/draft_export/__test_draft_export__/test_draft_export#link-tree/torch/nn/modules/module.py, lineno 1747, in _call_impl
        File /data/users/angelayi/fbsource/buck-out/v2/gen/fbcode/1beb9df83fd74b9a/scripts/angelayi/draft_export/__test_draft_export__/test_draft_export#link-tree/scripts/angelayi/draft_export/test_draft_export.py, lineno 138, in forward.
    Because of this, we have modified the dynamic shapes structure to be the following:
    ```
    dynamic_shapes = {'a': {0: 3}}
    ```
```

The result of this diff is also that `dynamic` logs are permanently turned on during draft export. Otherwise we cannot capture the `[guard added]` logs from symbolic_shapes.py.

Test Plan: `buck2 run @//mode/dev-nosan scripts/angelayi/draft_export:test_draft_export -- -r "test_shape_failure" `

Differential Revision: D64862374




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

ezyang reviewed (changes requested):
discussed in work chat. don't slam it into the regular TORCH_LOGS logs

ezyang reviewed (approved):
The global bool is funny business but ok. (nit: make it all caps?)

angelayi committed:
Add dtrace_structured logs for guard_added (#138748)

Summary:

For draft export, we would like to store the logs for some events that happen (like when a guard is added) to track where constraint violations happen. Since we want the logs to be structured, it would be nice to build off of the trace_structured logs. However since guard_added events happen a lot, we only want to log when we're in draft export for now.

So, we introduce a new logger `dtrace_structure` aka "detailed trace structured" or "debug trace structured" where we can add some more detailed logging like `guard_added` but won't clog up the normal `trace_structured` logs.

Test Plan: `buck2 run @//mode/dev-nosan scripts/angelayi/draft_export:test_draft_export -- -r "test_shape_failure" `

Reviewed By: ezyang

Differential Revision: D64862374

huydhn commented:
@pytorchbot merge -f 'A100 smoke test is not going to be done anytime soon due to queue'
